DICOM/src/Microsoft.Health.Anonymizer.Common/DateShiftFunction.cs

```diff
@@ -0,0 +1,79 @@
+// Licensed under the MIT License (MIT). See LICENSE in the repo root for license information.
+// -------------------------------------------------------------------------------------------------
+
+using System;
+using System.Text;
+using EnsureThat;
+using Microsoft.Health.Anonymizer.Common.Exceptions;
+using Microsoft.Health.Anonymizer.Common.Models;
+using Microsoft.Health.Anonymizer.Common.Settings;
+
+namespace Microsoft.Health.Anonymizer.Common
+{
+    public class DateShiftFunction
+    {
+        private const int DateShiftSeed = 131;
+        private readonly DateShiftSetting _dateShiftSetting;
+
+        public DateShiftFunction(DateShiftSetting dateShiftSetting)
+        {
+            EnsureArg.IsNotNull(dateShiftSetting, nameof(dateShiftSetting));
+
+            _dateShiftSetting = dateShiftSetting;
+        }
+
+        public string Shift(string inputString, AnonymizerValueTypes valueType)
+        {
+            EnsureArg.IsNotNull(inputString, nameof(inputString));
+
+            return valueType switch
+            {
+                AnonymizerValueTypes.Date => ShiftDateTime(inputString, DateTimeGlobalSettings.DateFormat),
+                AnonymizerValueTypes.DateTime => ShiftDateTime(inputString, DateTimeGlobalSettings.DateTimeFormat),
+                _ => throw new AnonymizerException(AnonymizerErrorCode.DateShiftFailed, "Unsupported value type. DateShift only applicable on Date or DateTime."),
+            };
+        }
+
+        public DateTimeOffset Shift(DateTimeOffset dateTime)
+        {
+            EnsureArg.IsNotNull<DateTimeOffset>(dateTime, nameof(dateTime));
+
+            try
+            {
+                DateTimeOffset newDateTime = new DateTimeOffset(dateTime.Year, dateTime.Month, dateTime.Day, 0, 0, 0, dateTime.Offset);
+                return newDateTime.AddDays(GetDateShiftValue());
+            }
+            catch (Exception ex)
+            {
+                throw new AnonymizerException(AnonymizerErrorCode.DateShiftFailed, "Failed to shift date.", ex);
+            }
+        }
+
+        private string ShiftDateTime(string inputString, string dateTimeFormat)
+        {
+            DateTimeOffset date = DateTimeUtility.ParseDateTimeString(inputString, dateTimeFormat);
+            return Shift(date).ToString(dateTimeFormat);
+        }
+
+        private int GetDateShiftValue()
+        {
+            int offset = 0;
+            var bytes = Encoding.UTF8.GetBytes(_dateShiftSetting.DateShiftKeyPrefix + _dateShiftSetting.DateShiftKey);
+            foreach (byte b in bytes)
+            {
+                offset = (int)(((offset * DateShiftSeed) + b) % ((2 * _dateShiftSetting.DateShiftRange) + 1));
+            }
+
+            offset -= (int)_dateShiftSetting.DateShiftRange;
+
+            return offset;
+        }
+
+        public void SetDateShiftPrefix(string value)
+        {
+            _dateShiftSetting.DateShiftKeyPrefix = value ?? string.Empty;
```

Comment:
🟡 SUGGESTION
In most XxxFunctions, we simply binds a setting to the XxxFunction and don't need to change the setting when function executes.
I know dateshift may have different prefix per resource, patient, etc. But do we need to share a Function with different settings?